Implemented arithmetic-instructions

jaborn opened the pull request:
I implemented some of the arithmetic instructions:
- add/ addu/ addu2/ addu4/ addu8/ addu16
- sub/ subu
- cmp/cmpu
- div
- mul/ mulu
-neg/ negu
Implementing for example "mulu" I had some problems because a value need to be stored in a special register. Therefor I want to use "put" but I am not sure of how to commit the value as the value refers to the operand y and z. So please have especially a look at this.

Commit:
Implemented arithmetic-instructions

Commit:
Bugfix

jonas-schievink commented:
May I ask how you test the implemented instructions?

jaborn committed:
Remove trailing whitespace

jaborn commented:
We talked about adding tests on the instructions but we only discussed it once at the beginning. I guess this would be very useful to do it that way so that we know that the functionality is correct.

jonas-schievink commented:
When I implement an emulator I always make sure to grab a runnable ROM right at the beginning and try to make it run. That way, I implement the instructions needed by the program (and skip the more obscure ones until later). I highly recommend that (or a similar) approach, since *nothing* is harder to debug than a wrong opcode implementation (this approach doesn't prevent them, but it's much better than doing no testing).

It also feels much more rewarding when some tracing capabilities are built into the interpreter (printing the current opcode or its semi-disassembled form, as well as parts of the current CPU state). You can literally see the program and CPU *do stuff*.

BarbaraButz reviewed (changes requested):
You need to remove some trailing whitespace to content Travis.
`>>> Seaching for lines with trailing whitespaces...	Error.
Found:	./src/machine/behavior/is/arithmetic/cmp.rs
Found:	./src/machine/behavior/is/arithmetic/neg.rs`

teiesti commented:
> Implementing for example "mulu" I had some problems because a value need to be stored in a special register. Therefor I want to use "put" but I am not sure of how to commit the value as the value refers to the operand y and z.

If you perform a multiplication, the result can have twice the size of the operands (because 2^64 * 2^64 == 2^(64+64) == 2^(2*64) == 2^128). 128 bit do not fit into a 64 bit-register, therefore MUL throws the upper half away. But MULU stores the upper half in the register rH ("himult").

What you have to do is:

1. Fetch the operands from $Y and $Z, as always.
2. Calculate the lower half.
3. Calculate the upper half. That might be a bit tricky since there is no operation to do that. You have to do some bit shifting.
4. Store the lower half in $X.
5. Store the upper half in rH.

The special register rH can be accessed in the following manner:
```rust
state.sr[R::H] = /*<some rvalue>*/;
```

About using PUT: Please don't do that! PUT needs to check if the special register is accessible. And PUT adds an unnecessary level of indirection.

> We talked about adding tests on the instructions but we only discussed it once at the beginning. I guess this would be very useful to do it that way so that we know that the functionality is correct.

The idea is, that every "instruction file" (like e.g. src/machine/behavior/is/arithmetic/add.rs) includes a test module that looks like this:

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn some_test() {
        // do some test
    }

    // do more tests

}
```

teiesti commented:
FYI: [There are efforts to include a u128 into Rust.][10] That would make our life much easier here. But as long as this feature is not available we have to emulate the behavior...

[10]: https://github.com/rust-lang/rfcs/blob/master/text/1504-int128.md

teiesti commented on src/machine/behavior/is/arithmetic/add.rs:
```diff
@@ -1,6 +1,11 @@
+    let op1: i64 = state.gpr[y].into();
+    let op2: i64 = state.gpr[z].into();
+    // execute
+    let res = op1 + op2;
```
What happens to the overflow?

teiesti commented on src/machine/behavior/is/arithmetic/addu.rs:
```diff
@@ -1,6 +1,11 @@
+    //  load operands
+    let op1: u64 = state.gpr[y].into();
+    let op2: u64 = state.gpr[z].into();
+    // execute
```
It would be nice if these comments start with a capital letter (since they are some kind of heading). It would also be nice if the "code blocks" have trailing newlines. This would bring more structure. Example:

```rust
pub fn addu(state: &mut State, x: u8, y: u8, z:u8) {
    // Load operands
    let op1: u64 = state.gpr[y].into();
    let op2: u64 = state.gpr[z].into();

    // Execute
    let res = op1.wrapping_add(op2);

    // Store results
    state.gpr[x] = res.into();
}
```

jaborn replied:
I will change most of them along with other corrections and I won't make a special commit for it. I hope this is ok.

teiesti commented on src/machine/behavior/is/arithmetic/cmp.rs:
```diff
@@ -1,6 +1,16 @@
+    let op2: i64 = state.gpr[z].into();
+    // check if op1 is greater, less or equal than op2
+    // store result 
+    if op1 < op2 {
```
Try this:

```rust
// Load operands
let op1: i64 = state.gpr[y].into();
let op2: i64 = state.gpr[z].into();

// Execute
let res: i64 = op1.cmp(&op2) as i64;

// Store results
state.gpr[x] = res.into();
```

See [here][10] and [here][20] and [here][30] to understand why that works.

[10]: https://doc.rust-lang.org/src/core/cmp.rs.html#200-210
[20]: https://doc.rust-lang.org/book/casting-between-types.html
[30]: http://rustbyexample.com/custom_types/enum/c_like.html

jonas-schievink replied:
Wow that's neat!

teiesti commented on src/machine/behavior/is/arithmetic/cmpu.rs:
```diff
@@ -1,6 +1,17 @@
+	// execute
+    let dif = op1.wrapping_sub(op2);
+    // store result
+    if dif > 0 {
```
--> CMP

teiesti commented on src/machine/behavior/is/arithmetic/div.rs:
```diff
@@ -1,6 +1,19 @@
+    let op2: i64 = state.gpr[z].into();
+    // check if denominator == 0
+    if op2 == 0 {
+    	put(state, (op1 as u8), 6, 0);
```
Please do not use PUT here. Details are explained in my long comment.

teiesti commented on src/machine/behavior/is/arithmetic/mul.rs:
```diff
@@ -1,6 +1,11 @@
+    let op1: i64 = state.gpr[y].into();
+    let op2: i64 = state.gpr[z].into();
+    // execute
+    let res = op1 * op2;
```
Panics on overflow (in debug mode). Should wrap (if I understand it right).

teiesti commented on src/machine/behavior/is/arithmetic/mulu.rs:
```diff
@@ -1,6 +1,18 @@
+    	state.gpr[x] = res.0.into();
+    } else {
+    	state.gpr[x] = res.0.into();
+    	let rh: u8 = (2 as u8).pow(64).wrapping_sub(op1 as u8);
```
I don't think, this is right!

teiesti commented on src/machine/behavior/is/arithmetic/neg.rs:
```diff
@@ -1,6 +1,10 @@
+    // load operand
+    let op2: i64 = state.gpr[z].into();
+    // execute
+    let res = (y as i64) - op2;
```
What happens to the overflow?

teiesti commented on src/machine/behavior/is/arithmetic/sub.rs:
```diff
@@ -1,6 +1,11 @@
+    let op1: i64 = state.gpr[y].into();
+    let op2: i64 = state.gpr[z].into();
+    // execute
+    let res = op1 - op2;
```
What happens to the overflow?

jaborn committed:
Remove trailing whitespaces

jaborn committed:
Remove tabs

jaborn committed:
Change the instruction 'cmp'

jaborn committed:
Change the instruction 'cmpu'

jaborn committed:
Use wrapping operations

jaborn committed:
Store the remainder without using 'put'

jaborn committed:
Rework 'mulu' (calculation of upper half)

jaborn committed:
Improve style

teiesti commented on src/machine/behavior/is/arithmetic/div.rs:
```diff
@@ -1,6 +1,21 @@
+    // check if denominator == 0
+    if op2 == 0 {
+        state.sr[R::R] = op1.into();
+        state.gpr[x] = (0 as i64).into();
```
Try this:

```rust
state.gpr[x] = 0i64.into();
```

teiesti commented on src/machine/behavior/is/arithmetic/mulu.rs:
```diff
@@ -1,6 +1,21 @@
+    // upper half
+    let upp_op1 = op1.wrapping_shl(32);
+    let upp_op2 = op2.wrapping_shl(32);
+    let upp_res = upp_op1.wrapping_mul(upp_op2);
```
I still don't think that is right. I will do the theory a little later.

jonas-schievink replied:
I suggest you just skip MULU for now, it's probably not used widely.

jaborn replied:
Ok. I will skip 'divu' as well because it is the same thing and work on it at the end.

jaborn committed:
Change casting

jaborn committed:
'Mulu' unimplemented

teiesti committed:
MULU implemented

teiesti commented:
Why so fast with MULU?

jaborn committed:
Add further arithmetic instructions
- addi/ addui/ addu2i/ addu4i/ addu8i/ addu16i
- cmpi/ cmpui
- divi
- muli/ mului
- negi/ negui
- subi/ subui

teiesti commented on src/machine/behavior/is/arithmetic/mului.rs:
```diff
@@ -1,6 +1,29 @@
+pub fn mului(state: &mut State, x: u8, y: u8, z: u8) {
+    // Load operand
+    let op1: u64 = state.gpr[y].into();
+
```
A reference to the theory in the other file would be nice!

jaborn committed:
Remove tabs

jaborn committed:
Add reference to theory in other file

jaborn committed:
Add further arithmetic instructions
- sl/sli
- slu/slui
- sr/sri
- sru/srui

teiesti committed:
Polished instruction DIV

teiesti committed:
Polished instruction DIVI

teiesti committed:
Polished instruction MULUI.

teiesti commented:
I have rebased this branch to be able to merge. And I have made some last polishing changes. This is the last call for reviews: If there is no issue until tomorrow morning, I will merge this pull-request... ;-)
